### PR TITLE
Documentation for various controllers and and lib utils

### PR DESCRIPTION
**Pull request by danjm**
Adds documentation app/scripts/notice-controller.js
app/scripts/ui.js
app/scripts/controllers/address-book.js
app/scripts/controllers/currency.js
app/scripts/controllers/preferences.js
app/scripts/controllers/shapeshift.js
app/scripts/lib/buy-eth-url.js
app/scripts/lib/environment-type.js
app/scripts/lib/get-first-preferred-lang-code.js
app/scripts/lib/is-popup-or-notification.js
app/scripts/lib/nodeify.js
app/scripts/lib/util.js

#3941

### CODE REVIEW

**Review by whymarrh — CHANGES_REQUESTED** [review]

File: app/scripts/controllers/shapeshift.js [hunk 1]
```diff
@@ -14,21 +25,54 @@ class ShapeshiftController {
     this.pollForUpdates()
   }
 
+  /**
+   * Represents, and contains data about, a single shapeshift transaction.
+   * @typedef {Object} ShapeShiftTx
+   * @property {string} depositAddress - An address at which to send a crypto deposit, so that eth can be sent to the
+   * user's Metamask account
+   * @property {string} depositType - An abbreviation of the type of crypto currency to be deposited.
+   * @constant {string} key - The 'shapeshift' key differentiates this from other types of txs in Metamask
```
[the remark this answers is not in the document]
Comment (danjm) [reply app/scripts/controllers/shapeshift.js]: Is it incorrect to use `@constant`? The issue is that this property will always be the string `'shapeshift'`, and I think that should be indicated in some way.

**Comment by danjm** [comment]
@whymarrh I believe I address all your comments in 23acddf, cc7e714 and e9ca7199

File: app/scripts/controllers/shapeshift.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (whymarrh) [reply app/scripts/controllers/shapeshift.js]: From reading the docs, I think this should be `@property`. It also seems that's what tooling recognizes (TS and my IDE at least)

**Comment by whymarrh** [comment]
@danjm outside of that last comment above this LGTM

**Comment by whymarrh** [comment]
One other thing I might comment on is that some of the docblocks added have introduced mixed tabs and spaces (e.g. `currency.js`) so it might be worth updating those from now

File: app/scripts/controllers/shapeshift.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (danjm) [reply app/scripts/controllers/shapeshift.js]: Fixed in 6d96b1a2

**Comment by danjm** [comment]
> One other thing I might comment on is that some of the docblocks added have introduced mixed tabs and spaces (e.g. currency.js) so it might be worth updating those from now

Fixed in e80bd230

**Review by whymarrh — APPROVED** [review]
LGTM

**Pull request merged**